---
title: "2021-01-06_atacseq"
output: html_document
---

```{r setup, include=FALSE}
#devtools::install_github("GreenleafLab/chromVARmotifs")
#source("http://bioconductor.org/biocLite.R")
#biocLite("chromVAR")
#biocLite("motifmatchr")
#biocLite("BSgenome.Hsapiens.UCSC.hg38")
#BiocManager::install("chromVAR")
#install.packages("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/chlsab_bsgenome/BSgenome.CSabaeus.UCSC.chlSab2_1.0.0.tar.gz", repos = NULL, type="source")
#BiocManager::install("motifmatchr")

library(chromVAR)
library(motifmatchr)
library(BSgenome.CSabaeus.UCSC.chlSab2)

library(plyr)
library(DESeq2)
library(ggplot2)
library(data.table)
library(biomaRt)
library(ggrepel)
library(pheatmap)
library(dplyr)
library(clusterProfiler)
library(stringr)
library(reactome.db)
library(OmnipathR)
library(gprofiler2)
library(igraph)
library(preprocessCore)
library(apeglm)
library(limma)

### helper function ####
substrLeft <- function(x, n){
  substr(x, 1, nchar(x)-n)
}

substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}
########################
```

```{r, eval=T}
# module load gcc/6.2.0 bedtools/2.27.1
# bedtools intersect -wa -a vero_sorted_peaks.clipped.nonzero.bed -b vero_consensus_recombinant.bed > vero_sorted_peaks.clipped.nonzero.recombinant.bed

peakfile <- "datafiles/vero_sorted_peaks.clipped.nonzero.xf_consensus.bed" # FDR 1% from macs2
bamfiles <- list.files(path="/Users/jonathanlee/Desktop/atacseq_vero_xf_screen/",pattern=".bam", full.names=T)
load("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/motifs/csab_cisbp_pwms.rda")
```

```{r, eval=T}
# read in ATAC-seq read pileups within peaks from bash
atac_counts <- read.table(file="datafiles/vero_atacseq_xf_counts.filt.txt", header=T)
atac_counts <- atac_counts[!duplicated(atac_counts),]
row.names(atac_counts) <- atac_counts[,1]
atac_counts <- atac_counts[,-1]
atac_counts <- atac_counts[,-c(13,14,20,21)]

# read in ATAC-seq library information
sampleinfo <- read.csv(file="sample_info.csv")
row.names(sampleinfo) <- paste0(sampleinfo$Batch, "-", sampleinfo$Plasmid, "-", sampleinfo$Gene, "-", sampleinfo$Replicate)

colnames(atac_counts) <- str_replace_all(unlist(strsplit(colnames(atac_counts), "_"))[c(F,F,T)], "S", "")
sampleinfo <- sampleinfo[match(colnames(atac_counts), sampleinfo$Number),]
colnames(atac_counts) <- row.names(sampleinfo)

sampleinfo$Group <- sampleinfo$Gene
sampleinfo$Group[which(sampleinfo$Gene %in% c("EV", "EGFP"))] <- "Ctrl"
sampleinfo$Group <- paste0(sampleinfo$Batch, "-", sampleinfo$Group)

sampleinfo$Group2 <- paste0(sampleinfo$Batch, "-", sampleinfo$Gene)
```

```{r, eval=T, fig.width=4, fig.height=6}
f <- list.files(path="datafiles/", full.names=T, pattern="plasmid_metrics.txt")[1]
plasmid_counts <- read.table(file=f, sep='\t', header=F)[1:34,c(2,1)]

for(f in list.files(path="datafiles/", full.names=T, pattern="plasmid_metrics.txt")[-1]){
  plasmid_counts <- merge(plasmid_counts, read.table(file=f, sep='\t', header=F)[1:34,], by.x=1, by.y=2)
}

colnames(plasmid_counts) <- c("plasmid", row.names(sampleinfo))
plasmid_counts <- plasmid_counts[-1,]

row.names(plasmid_counts) <- c(paste0("CMV.", c("EGFP", "EV", "nsp3", "nsp6", "S")), 
  paste0("EF1a.", c("E", "EGFP", "EV", "M", "N", "nsp1", "nsp10", "nsp11", "nsp12", "nsp13", "nsp14", "nsp15", "nsp16", "nsp2", "nsp4", "nsp5", "nsp7", "nsp8", "nsp9", "orf10", "orf3a", "orf3b", "orf6", "orf7a", "orf7b", "orf8", "orf9b", "orf9c"))) 
plasmid_counts <- plasmid_counts[,-1]

pheatmap(t(plasmid_counts), scale="row")
```

```{r, eval=T, fig.width=1, fig.height=0.8}
for(f in list.files(path="datafiles/", pattern="insertsize2.txt", full.names = T)){
  insertsize <- read.table(file=f, sep='\t', header=T)
  print(ggplot(insertsize, aes(x=size, y=count)) + geom_line() + theme_classic() + ylim(c(0,20000)) + xlim(c(0,1000)) + xlab("Insert Size (bp)") + ylab("# Reads") + ggtitle(subset(sampleinfo, Number==str_replace_all(strsplit(f, "_")[[1]][3], "S", ""))$Sample) + theme(plot.title = element_text(size = 10)))
}
```

```{r, eval=T, fig.width=1, fig.height=1}
tss_values <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="distToTSS.txt")){
  tss <- read.table(file=f, sep=' ', header=T)
  print(ggplot(tss, aes(x=Position/1000, y=Count)) + geom_line() + theme_classic() + ylim(c(0,2000)) + xlab("TSS Position (kb)") + ylab("# Reads")+ ggtitle(subset(sampleinfo, Number==str_replace_all(strsplit(f, "_")[[1]][3], "S", ""))$Sample) + theme(plot.title = element_text(size = 10))) 
  tss_values <- rbind(tss_values, cbind(sample=f, tss=mean(tss[c(2001 + seq(-5,5)),1])/mean(tss[1:200,1])))
}
```

```{r, eval=T}
tss_values <- data.frame(tss_values)
tss_values$TSS <- as.numeric(as.character(tss_values$tss))
tss_values$index <- str_replace_all(unlist(strsplit(tss_values$sample, "_"))[c(F,F,T,F)], "S", "")
tss_values <- merge(tss_values, sampleinfo[,c("Number", "Sample", "Group", "Group2")], by.x="index", by.y="Number")
tss_values <- tss_values[,c("Sample", "Group", "Group2", "TSS")]

```

```{r, eval=T, fig.width=3, fig.height=1.5}
ggplot(tss_values, aes(x=Group2, y=TSS)) + geom_point() + theme_classic() + geom_hline(yintercept=0, lty=3) + xlab("") + ylab("TSS-to-Background\nInsertion Ratio") + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```

```{r, eval=T, fig.width=1.5, fig.height=4}
nonmito_prop <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="mito.log")){
#for(f in list.files(path=".", pattern="mito.log")){
  nonmito_prop <- rbind(nonmito_prop, cbind(sample=f, prop=as.character(tail(read.table(file=f),1)$V5)))
}

nonmito_prop <- data.frame(nonmito_prop)
nonmito_prop$prop <- as.numeric(as.character(nonmito_prop$prop))

nonmito_prop$index <- str_replace_all(unlist(strsplit(nonmito_prop$sample, "_"))[c(F,F,T,F)], "S", "")
nonmito_prop <- merge(nonmito_prop, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
nonmito_prop <- nonmito_prop[,c("Sample", "Group", "prop")]

ggplot(nonmito_prop, aes(x=Sample, y=prop)) + geom_point() + ylim(c(0,1.1)) + theme_classic() + xlab("") + ylab("Non-Mitochondrial\nRead Proportion") + coord_flip()
```

```{r, eval=T, fig.width=2, fig.height=4}
peakanno <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="5000000.annstats.txt")){
#for(f in list.files(path=".", pattern="5000000.annstats.txt")){
  peakanno <- rbind(peakanno, cbind(sample=f, anno=as.character(head(read.table(file=f, sep='\t', header=T),5)[,1]), ratio=as.numeric(as.character(head(read.table(file=f, sep='\t', header=T),5)[,4]))))
}

peakanno <- data.frame(peakanno)
peakanno$ratio <- as.numeric(as.character(peakanno$ratio))

peakanno$index <- str_replace_all(unlist(strsplit(peakanno$sample, "_"))[c(F,F,T,F)], "S", "")
peakanno <- merge(peakanno, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
peakanno <- peakanno[,c("Sample", "Group", "anno", "ratio")]

ggplot(peakanno, aes(x=anno, y=ratio, fill=Sample)) + geom_bar(stat="identity", position="dodge") + theme_classic() + ylab("Log2 Fold Change") + xlab("Annotation") + geom_hline(yintercept=0, lty=3) + coord_flip() + theme(legend.position="none")
```

```{r, eval=T}
atac_dds <- DESeqDataSetFromMatrix(countData = atac_counts, colData = sampleinfo, design = ~Group)
#atac_dds <- estimateSizeFactors(atac_dds, controlGenes=row.names(atac_counts) %in% atac_invariant_tss)
atac_dds <- estimateSizeFactors(atac_dds)

atac_dds_B <- atac_dds
atac_dds_B$Group <- relevel(atac_dds_B$Group, "B-Ctrl")
```

```{r, eval=T}
atac_vst <- assay(varianceStabilizingTransformation(atac_dds))
```

```{r, eval=T, fig.width=2, fig.height=1.5}
library(viridis)

atac_pc <- prcomp(t(atac_vst))
```

```{r, eval=T, fig.width=2.5, fig.height=2}
for(b in unique(sampleinfo$Batch)){
atac_pc <- prcomp(t(atac_vst[,row.names(subset(sampleinfo, Batch==b))]))

print(ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst[,row.names(subset(sampleinfo, Batch==b))]),"Group2"]), aes(PC1, PC2)) + 
  geom_point(aes(colour = Sample), size=3)  + theme_classic() + xlab(paste0("PC1 (", 100*summary(atac_pc)$importance[2,1], "% of variance)")) + ylab(paste0("PC2 (", 100*summary(atac_pc)$importance[2,2], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red"))
}
```

```{r, eval=T}
atac_dds <- DESeq(atac_dds)
```

```{r, eval=T}
atac_sigpeaks <- data.frame()
for(b in unique(sampleinfo$Batch)){
for(l in unique(subset(sampleinfo, Batch==b & !(Gene %in% c("EV", "EGFP")))$Group)){
  atac_results_temp <- data.frame(results(atac_dds, contrast=c("Group", l, paste0(b,"-","Ctrl"))))
  atac_sigpeaks <- rbind(atac_sigpeaks, data.frame(group = l,
        sigpeaks = -dim(subset(atac_results_temp, padj < 0.05 & log2FoldChange < 0))[1],
        sign = "neg"))
  atac_sigpeaks <- rbind(atac_sigpeaks, data.frame(group = l,
        sigpeaks = dim(subset(atac_results_temp, padj < 0.05 & log2FoldChange > 0))[1],
        sign = "pos"))
}}

atac_sigpeaks$group2 <- unlist(strsplit(atac_sigpeaks$group, "-"))[c(F,T)]
atac_sigpeaks$group2 <- factor(atac_sigpeaks$group2, levels=c(paste0("nsp", 1:16), "S", "orf3a", "orf3b", "E", "M", "orf6", "orf7a", "orf7b", "orf8", "orf9b", "orf9c", "orf10", "N"))
```

```{r, eval=T, fig.width=2, fig.height=1}
ggplot(atac_sigpeaks, aes(x=group2, y=sigpeaks, color=sign, pch=sign)) + geom_point() + ylim(c(-3000, 3000)) + theme_classic() + geom_hline(yintercept=0) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant Peaks\n(vs. Control)") + xlab("") + scale_color_manual(values=c("red", "black")) + scale_shape_manual(values=c(1, 2))
```

```{r, eval=T}
atac_dds_B <- DESeq(atac_dds_B)
```

```{r, eval=T}
results_vero_spike <- data.frame(lfcShrink(atac_dds_B, coef=10))
results_vero_nsp1 <- data.frame(lfcShrink(atac_dds, coef=2))
```

```{r, eval=T}
atacseq_vero_sars2.mock <- read.table(file="../atacseq_vero_virus/vero_sars2.mock_atacseq.txt", sep='\t', header=T)
```

```{r, eval=T}
cor((subset(merge(atacseq_vero_sars2.mock, results_vero_spike, by=0), padj.x < 0.2 & padj.y < 0.2)$log2FoldChange.x),
    (subset(merge(atacseq_vero_sars2.mock, results_vero_spike, by=0), padj.x < 0.2 & padj.y < 0.2)$log2FoldChange.y))

cor((subset(merge(atacseq_vero_sars2.mock, results_vero_nsp1, by=0), padj.x < 0.2 & padj.y < 0.2)$log2FoldChange.x),
    (subset(merge(atacseq_vero_sars2.mock, results_vero_nsp1, by=0), padj.x < 0.2 & padj.y < 0.2)$log2FoldChange.y))

cor((subset(merge(results_vero_spike, results_vero_nsp1, by=0), padj.x < 0.2 & padj.y < 0.2)$log2FoldChange.x),
    (subset(merge(results_vero_spike, results_vero_nsp1, by=0), padj.x < 0.2 & padj.y < 0.2)$log2FoldChange.y))
```

```{r, eval=T, fig.width=1.5, fig.height=1.5}
ggplot(subset(merge(atacseq_vero_sars2.mock, results_vero_spike, by=0), padj.x < 0.2 & padj.y < 0.2)) + geom_point(aes(x=(log2FoldChange.x), y=(log2FoldChange.y)), alpha=0.5) + xlim(c(-3.5,3.5)) + ylim(c(-3.5,3.5)) + geom_smooth(method="lm", formula=y~x, aes(x=(log2FoldChange.x), y=(log2FoldChange.y))) + theme_classic() + geom_hline(yintercept=0, lty=3) + geom_vline(xintercept=0, lty=3) + xlab("Vero SARS-CoV-2 vs. Mock\nLog2 Fold Change, ATAC-seq") + ylab("Vero Spike vs. Control\nLog2 Fold Change, ATAC-seq")

ggplot(subset(merge(atacseq_vero_sars2.mock, results_vero_nsp1, by=0), padj.x < 0.2 & padj.y < 0.2)) + geom_point(aes(x=(log2FoldChange.x), y=(log2FoldChange.y)), alpha=0.5) + xlim(c(-3.5,3.5)) + ylim(c(-3.5,3.5)) + geom_smooth(method="lm", formula=y~x, aes(x=(log2FoldChange.x), y=(log2FoldChange.y))) + theme_classic() + geom_hline(yintercept=0, lty=3) + geom_vline(xintercept=0, lty=3) + xlab("Vero SARS-CoV-2 vs. Mock\nLog2 Fold Change, ATAC-seq") + ylab("Vero Nsp1 vs. Control\nLog2 Fold Change, ATAC-seq")
```

```{r, eval=T}
write.table(results_vero_nsp1, file="atacseq_vero_nsp1.txt", sep='\t', quote=F)
write.table(results_vero_spike, file="atacseq_vero_spike.txt", sep='\t', quote=F)
```


