---
title: "2021-01-08_atacseq"
output: html_document
---

```{r setup, include=FALSE}
#devtools::install_github("GreenleafLab/chromVARmotifs")
#source("http://bioconductor.org/biocLite.R")
#biocLite("chromVAR")
#biocLite("motifmatchr")
#biocLite("BSgenome.Hsapiens.UCSC.hg38")
#BiocManager::install("chromVAR")
#install.packages("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/chlsab_bsgenome/BSgenome.CSabaeus.UCSC.chlSab2_1.0.0.tar.gz", repos = NULL, type="source")
#BiocManager::install("motifmatchr")

library(chromVAR)
library(motifmatchr)
library(BSgenome.CSabaeus.UCSC.chlSab2)

library(plyr)
library(DESeq2)
library(ggplot2)
library(data.table)
library(biomaRt)
library(ggrepel)
library(pheatmap)
library(dplyr)
library(clusterProfiler)
library(stringr)
library(reactome.db)
library(OmnipathR)
library(gprofiler2)
library(igraph)
library(preprocessCore)
library(apeglm)
library(limma)

### helper function ####
substrLeft <- function(x, n){
  substr(x, 1, nchar(x)-n)
}

substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}
########################
```

```{r, eval=T}
# module load gcc/6.2.0 bedtools/2.27.1
# bedtools intersect -wa -a vero_sorted_peaks.clipped.nonzero.bed -b vero_consensus_xf-spike-cov.bed > vero_sorted_peaks.clipped.nonzero.xf-spike-cov.bed

peakfile <- "datafiles/vero_sorted_peaks.clipped.nonzero.xf-spike-cov.bed" # FDR 1% from macs2
bamfiles <- list.files(path="/Users/jonathanlee/Desktop/atacseq_vero_spike-cov_xf/",pattern=".bam", full.names=T)
load("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/motifs/csab_cisbp_pwms.rda")
```

```{r, eval=T}
# read in ATAC-seq read pileups within peaks from bash
atac_counts <- read.table(file="datafiles/vero_atacseq_xf-spike-cov_count.filt.txt", header=T)
atac_counts <- atac_counts[!duplicated(atac_counts),]
row.names(atac_counts) <- atac_counts[,1]
atac_counts <- atac_counts[,-1]

# read in ATAC-seq library information
sampleinfo <- read.csv(file="sample_info.csv")
row.names(sampleinfo) <- sampleinfo$Sample

colnames(atac_counts) <- str_replace_all(unlist(strsplit(colnames(atac_counts), "_"))[c(F,F,T)], "S", "")
sampleinfo <- sampleinfo[match(colnames(atac_counts), sampleinfo$Number),]
colnames(atac_counts) <- row.names(sampleinfo)

sampleinfo$Group <- unlist(strsplit(str_replace_all(sampleinfo$Sample, "-Spike", ""), "-"))[c(F,F,T,F)]
sampleinfo$Group2 <- sampleinfo$Group
sampleinfo$Group2[which(sampleinfo$Group2 %in% c("EV", "EGFP"))] <- "Ctrl"
```

```{r, eval=T, fig.width=2.5, fig.height=2.5}
f <- list.files(path="datafiles/", full.names=T, pattern="plasmid_metrics.txt")[1]
plasmid_counts <- read.table(file=f, sep='\t', header=F)[10:19,c(2,1)]

for(f in list.files(path="datafiles/", full.names=T, pattern="plasmid_metrics.txt")[-1]){
  plasmid_counts <- merge(plasmid_counts, read.table(file=f, sep='\t', header=F)[10:19,], by.x=1, by.y=2)
}

colnames(plasmid_counts) <- c("plasmid", row.names(sampleinfo))

row.names(plasmid_counts) <- c("EV", "229E", "EGFP", "HKU1", "MERS", "NL63", "OC43", "SARS1", "SARS2-S2P", "SARS2") 
plasmid_counts <- plasmid_counts[,-1]

pheatmap((plasmid_counts), scale="column")
```

```{r, eval=T, fig.width=1, fig.height=0.8}
for(f in list.files(path="datafiles/", pattern="insertsize2.txt", full.names = T)){
  insertsize <- read.table(file=f, sep='\t', header=T)
  print(ggplot(insertsize, aes(x=size, y=count)) + geom_line() + theme_classic() + ylim(c(0,20000)) + xlim(c(0,1000)) + xlab("Insert Size (bp)") + ylab("# Reads") + ggtitle(subset(sampleinfo, Number==str_replace_all(strsplit(f, "_")[[1]][3], "S", ""))$Sample) + theme(plot.title = element_text(size = 6)))
}
```

```{r, eval=T, fig.width=1, fig.height=1}
tss_values <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="distToTSS.txt")){
  tss <- read.table(file=f, sep=' ', header=T)
  print(ggplot(tss, aes(x=Position/1000, y=Count)) + geom_line() + theme_classic() + ylim(c(0,2000)) + xlab("TSS Position (kb)") + ylab("# Reads")+ ggtitle(subset(sampleinfo, Number==str_replace_all(strsplit(f, "_")[[1]][3], "S", ""))$Sample) + theme(plot.title = element_text(size = 6))) 
  tss_values <- rbind(tss_values, cbind(sample=f, tss=mean(tss[c(2001 + seq(-5,5)),1])/mean(tss[1:200,1])))
}
```

```{r, eval=T}
tss_values <- data.frame(tss_values)
tss_values$TSS <- as.numeric(as.character(tss_values$tss))
tss_values$index <- str_replace_all(unlist(strsplit(tss_values$sample, "_"))[c(F,F,T,F)], "S", "")
tss_values <- merge(tss_values, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
tss_values <- tss_values[,c("Sample", "Group", "TSS")]
tss_values$Group <- factor(tss_values$Group, levels=c("EV", "EGFP", "SARS1", "SARS2", "MERS", "HKU1", "NL63", "OC43", "229E", "SARS2S2P"))
```

```{r, eval=T, fig.width=1.2, fig.height=1.5}
#ggplot(tss_values, aes(x=Sample, y=TSS)) + geom_bar(stat="identity") + theme_classic() + geom_hline(yintercept=0) + coord_flip() + xlab("") + ylab("TSS-to-Background\nInsertion Ratio")

ggplot(tss_values, aes(x=Group, y=TSS)) + geom_point() + theme_classic() + geom_hline(yintercept=0, lty=3) + xlab("") + ylab("TSS-to-Background\nInsertion Ratio") + theme(axis.text.x = element_text(angle=90, hjust=1, vjust=0.5))
```

```{r, eval=T, fig.width=1.5, fig.height=2}
nonmito_prop <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="mito.log")){
#for(f in list.files(path=".", pattern="mito.log")){
  nonmito_prop <- rbind(nonmito_prop, cbind(sample=f, prop=as.character(tail(read.table(file=f),1)$V5)))
}

nonmito_prop <- data.frame(nonmito_prop)
nonmito_prop$prop <- as.numeric(as.character(nonmito_prop$prop))

nonmito_prop$index <- str_replace_all(unlist(strsplit(nonmito_prop$sample, "_"))[c(F,F,T,F)], "S", "")
nonmito_prop <- merge(nonmito_prop, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
nonmito_prop <- nonmito_prop[,c("Sample", "Group", "prop")]

ggplot(nonmito_prop, aes(x=Sample, y=prop)) + geom_point() + ylim(c(0,1.1)) + theme_classic() + xlab("") + ylab("Non-Mitochondrial\nRead Proportion") + coord_flip()
```

```{r, eval=T, fig.width=2, fig.height=4}
peakanno <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="5000000.annstats.txt")){
#for(f in list.files(path=".", pattern="5000000.annstats.txt")){
  peakanno <- rbind(peakanno, cbind(sample=f, anno=as.character(head(read.table(file=f, sep='\t', header=T),5)[,1]), ratio=as.numeric(as.character(head(read.table(file=f, sep='\t', header=T),5)[,4]))))
}

peakanno <- data.frame(peakanno)
peakanno$ratio <- as.numeric(as.character(peakanno$ratio))

peakanno$index <- str_replace_all(unlist(strsplit(peakanno$sample, "_"))[c(F,F,T,F)], "S", "")
peakanno <- merge(peakanno, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
peakanno <- peakanno[,c("Sample", "Group", "anno", "ratio")]

ggplot(peakanno, aes(x=anno, y=ratio, fill=Sample)) + geom_bar(stat="identity", position="dodge") + theme_classic() + ylab("Log2 Fold Change") + xlab("Annotation") + geom_hline(yintercept=0, lty=3) + coord_flip() + theme(legend.position="none")
```

```{r, eval=T}
atac_dds <- DESeqDataSetFromMatrix(countData = atac_counts, colData = sampleinfo, design = ~Group2)
atac_dds <- estimateSizeFactors(atac_dds)
```

```{r, eval=T}
atac_vst <- assay(varianceStabilizingTransformation(atac_dds))
```

```{r, eval=T, fig.width=2, fig.height=1.5}
library(viridis)

atac_pc <- prcomp(t(atac_vst))

ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst),"Group"]), aes(PC1, PC2)) + 
  geom_point(aes(colour = Sample, pch = Sample), size=2)  + theme_classic() + xlab(paste0("PC1 (", 100*summary(atac_pc)$importance[2,1], "% of variance)")) + ylab(paste0("PC2 (", 100*summary(atac_pc)$importance[2,2], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red") + 
  scale_colour_manual(values=viridis(4)[c(2,4,4,2,3,2,2,3,3,1)])+   
  scale_shape_manual(values=c(19, 17, 15, 18)[c(1,1,2,2,1,3,4,2,3,1)])

ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst),"Group"]), aes(PC3, PC4)) + 
  geom_point(aes(colour = Sample), size=3)  + theme_classic() + xlab(paste0("PC3 (", 100*summary(atac_pc)$importance[2,3], "% of variance)")) + ylab(paste0("PC4 (", 100*summary(atac_pc)$importance[2,4], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red")
```

```{r, eval=T, fig.width=2, fig.height=1.5}
atac_pc <- prcomp(t(atac_vst[,c(3:6, 9:10, 15:18)]))

ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst[,c(3:6, 9:10, 15:18)]),"Group"]), aes(PC1, PC2)) + 
  geom_point(aes(colour = Sample), size=3)  + theme_classic() + xlab(paste0("PC1 (", 100*summary(atac_pc)$importance[2,1], "% of variance)")) + ylab(paste0("PC2 (", 100*summary(atac_pc)$importance[2,2], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red")

ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst[,c(3:6, 9:10, 15:18)]),"Group"]), aes(PC3, PC4)) + 
  geom_point(aes(colour = Sample), size=3)  + theme_classic() + xlab(paste0("PC3 (", 100*summary(atac_pc)$importance[2,3], "% of variance)")) + ylab(paste0("PC4 (", 100*summary(atac_pc)$importance[2,4], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red")
```

```{r, eval=T, fig.width=2, fig.height=1.5}
library(Rtsne)
set.seed(1)
tsne_lr <- Rtsne(t(atac_vst[,c(3:6, 9:10, 15:20)]), dims = 2, perplexity = 3,  max_iter = 10000)
tsne_lr <- data.frame(colnames(atac_vst[,c(3:6, 9:10, 15:20)]), tsne_lr$Y, 
                      Sample=sampleinfo[colnames(atac_vst[,c(3:6, 9:10, 15:20)]),"Group"])
colnames(tsne_lr) <- c("Name", "dim1", "dim2", "Sample")

ggplot(tsne_lr, aes(dim1, dim2)) + geom_point(aes(color=Sample), size=3) + theme_classic() + xlab("TSNE Dimension 1") + ylab("TSNE Dimension 2") #+ theme(legend.position = "none")
```

```{r, eval=T, fig.width=2, fig.height=1.5}
library(Rtsne)
set.seed(1)
tsne_lr <- Rtsne(t(atac_vst), dims = 2, perplexity = 3,  max_iter = 10000)
tsne_lr <- data.frame(colnames(atac_vst), tsne_lr$Y, 
                      Sample=sampleinfo[colnames(atac_vst),"Group"])
colnames(tsne_lr) <- c("Name", "dim1", "dim2", "Sample")

ggplot(tsne_lr, aes(dim1, dim2)) + geom_point(aes(color=Sample), size=3) + theme_classic() + xlab("TSNE Dimension 1") + ylab("TSNE Dimension 2") #+ theme(legend.position = "none")
```


```{r, eval=T}
atac_dds <- DESeq(atac_dds)
```

```{r, eval=T}
atac_sigpeaks <- data.frame()
for(l in setdiff(sampleinfo$Group2, "Ctrl")){
  atac_sigpeaks <- rbind(atac_sigpeaks, data.frame(group = l,
        sigpeaks = -dim(subset(results(atac_dds, contrast=c("Group2", l, "Ctrl")), padj < 0.05 & log2FoldChange < 0))[1],
        sign = "neg"))
  atac_sigpeaks <- rbind(atac_sigpeaks, data.frame(group = l,
        sigpeaks = dim(subset(results(atac_dds, contrast=c("Group2", l, "Ctrl")), padj < 0.05 & log2FoldChange > 0))[1],
        sign = "pos"))
}
```

```{r, eval=T, fig.width=1, fig.height=1.2}
atac_sigpeaks$group <- factor(atac_sigpeaks$group, levels=c("SARS1", "SARS2", "MERS", "HKU1", "NL63", "OC43", "229E", "SARS2S2P"))
ggplot(subset(atac_sigpeaks, group != "SARS2S2P"), aes(x=group, y=sigpeaks, color=sign, pch=sign)) + geom_point() + ylim(c(-5000, 5000)) + theme_classic() + geom_hline(yintercept=0) + scale_color_manual(values=c("red", "black")) + scale_shape_manual(values=c(1, 2)) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant Peaks\n(vs. Control)") + xlab("")
```

```{r, eval=T, fig.width=1, fig.height=1.2}
atac_sigpeaks$group <- factor(atac_sigpeaks$group, levels=c("SARS1", "SARS2", "MERS", "HKU1", "NL63", "OC43", "229E", "SARS2S2P"))
ggplot(subset(atac_sigpeaks), aes(x=group, y=sigpeaks, color=sign, pch=sign)) + geom_point() + ylim(c(-5000, 5000)) + theme_classic() + geom_hline(yintercept=0) + scale_color_manual(values=c("red", "black")) + scale_shape_manual(values=c(1, 2)) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant Peaks\n(vs. Control)") + xlab("")
```

```{r, eval=T, fig.width=0.8, fig.height=1.2}
atac_sigpeaks$group <- factor(atac_sigpeaks$group, levels=c("SARS1", "SARS2", "MERS", "HKU1", "NL63", "OC43", "229E", "SARS2S2P"))
ggplot(subset(atac_sigpeaks, group %in% c("SARS2", "SARS2S2P")), aes(x=group, y=sigpeaks, color=sign, pch=sign)) + geom_point() + ylim(c(-5000, 5000)) + theme_classic() + geom_hline(yintercept=0) + scale_color_manual(values=c("red", "black")) + scale_shape_manual(values=c(1, 2)) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant Peaks\n(vs. Control)") + xlab("")
```

```{r, eval=T, fig.width=1.5, fig.height=1}
atac_sigpeaks$group <- factor(atac_sigpeaks$group, levels=rev(c("SARS1", "SARS2", "MERS", "HKU1", "NL63", "OC43", "229E", "SARS2S2P")))

ggplot(subset(atac_sigpeaks, group != "SARS2S2P"), aes(x=group, y=sigpeaks, color=sign, pch=sign)) + geom_point() + ylim(c(-5000, 5000)) + theme_classic() + geom_hline(yintercept=0) + scale_color_manual(values=c("red", "black")) + scale_shape_manual(values=c(1, 2)) + theme(legend.position="none") + ylab("# Significant Peaks (vs. Control)") + xlab("") + coord_flip()
```

# fgsea - virus vs. spike only


```{r, eval=T}
results_sars1 <- results(atac_dds, contrast=c("Group2", "SARS1", "Ctrl"))
results_sars2 <- results(atac_dds, contrast=c("Group2", "SARS2", "Ctrl"))
results_mers <- results(atac_dds, contrast=c("Group2", "MERS", "Ctrl"))
```

```{r, eval=T}
atacseq_vero_sars1.mock <- read.table(file="../atacseq_vero_virus/vero_sars1.mock_atacseq.txt", sep='\t', header=T)
atacseq_vero_sars2.mock <- read.table(file="../atacseq_vero_virus/vero_sars2.mock_atacseq.txt", sep='\t', header=T)
atacseq_vero_mers.mock <- read.table(file="../atacseq_vero_virus/vero_mers.mock_atacseq.txt", sep='\t', header=T)
```

```{r, eval=T}
results_mat <- merge(merge(data.frame(results_sars1)[,c("log2FoldChange", "padj")], data.frame(results_sars2)[,c("log2FoldChange", "padj")], by=0), data.frame(results_mers)[,c("log2FoldChange", "padj")], by.x=1, by.y=0)

results_virus_mat <- merge(merge(data.frame(atacseq_vero_sars1.mock )[,c("log2FoldChange", "padj")], data.frame(atacseq_vero_sars2.mock )[,c("log2FoldChange", "padj")], by=0), data.frame(atacseq_vero_mers.mock )[,c("log2FoldChange", "padj")], by.x=1, by.y=0)

results_full_mat <- merge(results_mat, results_virus_mat, by=1)
colnames(results_full_mat) <- c("position", paste0(rep(c("log2FoldChange.", "padj."), 6), c(rep("sars1", 2), rep("sars2", 2), rep("mers", 2), rep("sars1.v", 2), rep("sars2.v", 2), rep("mers.v", 2))))
pheatmap(cor(results_full_mat[,c(F,T)]))
cor(results_full_mat[,c(F,T)])
```

```{r, eval=T}
results_sars2.sars1 <- results(atac_dds, contrast=c("Group2", "SARS2", "SARS1"))
results_sars2.mers <- results(atac_dds, contrast=c("Group2", "SARS2", "MERS"))
results_mers.sars1 <- results(atac_dds, contrast=c("Group2", "MERS", "SARS1"))
```

```{r, eval=T}
atac_diff_peaks <- list(sars1_up = row.names(subset(results_sars1, padj < 0.05 & log2FoldChange > 0)[order(subset(results_sars1, padj < 0.05 & log2FoldChange > 0)$padj),])[1:200],
     sars1_dn = row.names(subset(results_sars1, padj < 0.05 & log2FoldChange < 0)[order(subset(results_sars1, padj < 0.05 & log2FoldChange < 0)$padj),])[1:200],
     sars2_up = row.names(subset(results_sars2, padj < 0.05 & log2FoldChange > 0)[order(subset(results_sars2, padj < 0.05 & log2FoldChange > 0)$padj),])[1:200],
     sars2_dn = row.names(subset(results_sars2, padj < 0.05 & log2FoldChange < 0)[order(subset(results_sars2, padj < 0.05 & log2FoldChange < 0)$padj),])[1:200],
     mers_up = row.names(subset(results_mers, padj < 0.05 & log2FoldChange > 0)[order(subset(results_mers, padj < 0.05 & log2FoldChange > 0)$padj),])[1:200],
     mers_dn = row.names(subset(results_mers, padj < 0.05 & log2FoldChange < 0)[order(subset(results_mers, padj < 0.05 & log2FoldChange < 0)$padj),])[1:200])
```


```{r, eval=T}
atac_diff_peaks <- list(sars1_up = row.names(subset(results_sars1, padj < 0.05 & log2FoldChange > 0)[order(subset(results_sars1, padj < 0.05 & log2FoldChange > 0)$padj),])[1:500],
     sars1_dn = row.names(subset(results_sars1, padj < 0.05 & log2FoldChange < 0)[order(subset(results_sars1, padj < 0.05 & log2FoldChange < 0)$padj),])[1:500],
     sars2_up = row.names(subset(results_sars2, padj < 0.05 & log2FoldChange > 0)[order(subset(results_sars2, padj < 0.05 & log2FoldChange > 0)$padj),])[1:500],
     sars2_dn = row.names(subset(results_sars2, padj < 0.05 & log2FoldChange < 0)[order(subset(results_sars2, padj < 0.05 & log2FoldChange < 0)$padj),])[1:500],
     mers_up = row.names(subset(results_mers, padj < 0.05 & log2FoldChange > 0)[order(subset(results_mers, padj < 0.05 & log2FoldChange > 0)$padj),])[1:500],
     mers_dn = row.names(subset(results_mers, padj < 0.05 & log2FoldChange < 0)[order(subset(results_mers, padj < 0.05 & log2FoldChange < 0)$padj),])[1:500])
```


```{r, eval=T}
atacseq_vero_sars1.mock <- read.table(file="../atacseq_vero_virus/vero_sars1.mock_atacseq.txt", sep='\t', header=T)
atacseq_vero_sars2.mock <- read.table(file="../atacseq_vero_virus/vero_sars2.mock_atacseq.txt", sep='\t', header=T)
atacseq_vero_mers.mock <- read.table(file="../atacseq_vero_virus/vero_mers.mock_atacseq.txt", sep='\t', header=T)

vero_sars1.mock_ranks <- sign(subset(atacseq_vero_sars1.mock, padj < 0.2)$log2FoldChange) * -log10(subset(atacseq_vero_sars1.mock, padj < 0.2)$padj)
names(vero_sars1.mock_ranks) <- row.names(subset(atacseq_vero_sars1.mock, padj < 0.2))

vero_sars2.mock_ranks <- sign(subset(atacseq_vero_sars2.mock, padj < 0.2)$log2FoldChange) * -log10(subset(atacseq_vero_sars2.mock, padj < 0.2)$padj)
names(vero_sars2.mock_ranks) <- row.names(subset(atacseq_vero_sars2.mock, padj < 0.2))

vero_mers.mock_ranks <- sign(subset(atacseq_vero_mers.mock, padj < 0.2)$log2FoldChange) * -log10(subset(atacseq_vero_mers.mock, padj < 0.2)$padj)
names(vero_mers.mock_ranks) <- row.names(subset(atacseq_vero_mers.mock, padj < 0.2))

```

```{r, eval=T}
vero_sars1.mock_ranks <- subset(atacseq_vero_sars1.mock)$log2FoldChange
names(vero_sars1.mock_ranks) <- row.names(subset(atacseq_vero_sars1.mock))

vero_sars2.mock_ranks <- subset(atacseq_vero_sars2.mock)$log2FoldChange
names(vero_sars2.mock_ranks) <- row.names(subset(atacseq_vero_sars2.mock))

vero_mers.mock_ranks <- subset(atacseq_vero_mers.mock)$log2FoldChange
names(vero_mers.mock_ranks) <- row.names(subset(atacseq_vero_mers.mock))

```

```{r, eval=T}
vero_sars1.mock_ranks <- sign(subset(atacseq_vero_sars1.mock)$log2FoldChange) * -log10(subset(atacseq_vero_sars1.mock)$padj)
names(vero_sars1.mock_ranks) <- row.names(subset(atacseq_vero_sars1.mock))

vero_sars2.mock_ranks <- sign(subset(atacseq_vero_sars2.mock)$log2FoldChange) * -log10(subset(atacseq_vero_sars2.mock)$padj)
names(vero_sars2.mock_ranks) <- row.names(subset(atacseq_vero_sars2.mock))

vero_mers.mock_ranks <- sign(subset(atacseq_vero_mers.mock)$log2FoldChange) * -log10(subset(atacseq_vero_mers.mock)$padj)
names(vero_mers.mock_ranks) <- row.names(subset(atacseq_vero_mers.mock))

```

```{r, eval=T}
library(fgsea)
fgsea_sars1.mock <- fgsea(stats=vero_sars1.mock_ranks, pathways=atac_diff_peaks, eps=0, nPermSimple=1000)
fgsea_sars2.mock <- fgsea(stats=vero_sars2.mock_ranks, pathways=atac_diff_peaks, eps=0, nPermSimple=1000)
fgsea_mers.mock <- fgsea(stats=vero_mers.mock_ranks, pathways=atac_diff_peaks, eps=0, nPermSimple=1000)
```

```{r, eval=T}
fgsea_sars1.mock
fgsea_sars2.mock
fgsea_mers.mock
```


# ChromVAR analysis

```{r, eval=T}
peaks.bed <- read.table(file=peakfile, sep='\t')
peaks <- readNarrowpeaks(peakfile)
peaks@seqnames@values <- factor(str_replace_all(peaks@seqnames@values, "chr", ""))
peaks@seqinfo@seqnames <- str_replace_all(peaks@seqinfo@seqnames, "chr", "")

bam_index <- str_replace_all(str_replace_all(bamfiles, ".*S", ""), "_.*","")

fragment_counts <- getCounts(bamfiles, peaks, 
                              paired =  TRUE, 
                              by_rg = F, 
                              format = "bam", 
                              colData = DataFrame(row.names = sampleinfo[match(bam_index, sampleinfo$Number),]$Sample,
                                                  group = sampleinfo[match(bam_index, sampleinfo$Number),]$Group))

fragment_counts@rowRanges@seqnames@values <- factor(paste0("chr",fragment_counts@rowRanges@seqnames@values))
fragment_counts@rowRanges@seqnames@values <- factor(fragment_counts@rowRanges@seqnames@values, levels=as.character(fragment_counts@rowRanges@seqnames@values))
fragment_counts@rowRanges@seqinfo@seqnames <- paste0("chr",fragment_counts@rowRanges@seqinfo@seqnames)

fragment_counts <- addGCBias(fragment_counts, genome = BSgenome.CSabaeus.UCSC.chlSab2)
counts_filtered <- filterSamples(fragment_counts)
counts_filtered <- filterPeaks(counts_filtered) # min_in_peaks 10%

motif_ix <- matchMotifs(csab_cisbp_pwms, counts_filtered, genome = BSgenome.CSabaeus.UCSC.chlSab2)

# computing deviations
dev <- computeDeviations(object = counts_filtered, annotations = motif_ix)

dev <- dev[which(!is.na(rowSums(deviations(dev)))),]
```

```{r, eval=T, fig.width=1.7, fig.height=1.2}
dev_mat <- deviations(dev)
dev_mat <- dev_mat[which(!is.na(rowSums(dev_mat))),]
chromvar_pc <- prcomp(t(dev_mat), scale=T)

ggplot(data.frame(chromvar_pc$x, Sample=sampleinfo[match(bam_index, sampleinfo$Number),]$Sample,
                    Group = sampleinfo[match(bam_index, sampleinfo$Number),]$Group), aes(PC1, PC2)) + 
  geom_point(aes(colour = Group), size=4) + scale_color_brewer(type="qual", palette=1)  + theme_classic() + xlab(paste0("ChromVAR PC1\n(", format(round(100*summary(chromvar_pc)$importance[2,1], digits=1), nsmall=1), "% of variance)")) + ylab(paste0("ChromVAR PC2\n(", format(round(100*summary(chromvar_pc)$importance[2,2], digits=1), nsmall=1), "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red")
```

```{r, eval=T}
atac_sigmotifs <- data.frame()
for(l in setdiff(sampleinfo$Group2, "Ctrl")){
  atac_sigmotifs <- rbind(atac_sigmotifs, data.frame(group = l,
        sigmotifs = -sum(sign(rowMeans(deviations(dev)[,subset(sampleinfo, Group2==l)$Sample]) - rowMeans(deviations(dev)[,subset(sampleinfo, Group2=="Ctrl")$Sample])) == -1 & differentialDeviations(dev[,subset(sampleinfo, Group2%in% c("Ctrl", l))$Sample], groups="group")$p_value_adjusted < 0.05),
        sign = "neg"))
  atac_sigmotifs <- rbind(atac_sigmotifs, data.frame(group = l,
        sigmotifs = sum(sign(rowMeans(deviations(dev)[,subset(sampleinfo, Group2==l)$Sample]) - rowMeans(deviations(dev)[,subset(sampleinfo, Group2=="Ctrl")$Sample])) == 1 & differentialDeviations(dev[,subset(sampleinfo, Group2%in% c("Ctrl", l))$Sample], groups="group")$p_value_adjusted < 0.05),
        sign = "pos"))
}
```

```{r, eval=T, fig.width=1, fig.height=1.2}
ggplot(atac_sigmotifs, aes(x=group, y=sigmotifs, color=sign)) + geom_point() + ylim(c(-10, 10)) + theme_classic() + geom_hline(yintercept=0) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant TF Motifs\n(vs. Control)") + xlab("")
```

```{r, eval=T}
ggplot(data.frame(score=dev_mat["ENSCSAG00000011345",], sample=colnames(dev_mat))) + geom_point(aes(x=sample,y=score)) + coord_flip()
```



