---
title: "2021-01-07_atacseq"
output: html_document
---

```{r setup, include=FALSE}
#devtools::install_github("GreenleafLab/chromVARmotifs")
#source("http://bioconductor.org/biocLite.R")
#biocLite("chromVAR")
#biocLite("motifmatchr")
#biocLite("BSgenome.Hsapiens.UCSC.hg38")
#BiocManager::install("chromVAR")
#install.packages("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/chlsab_bsgenome/BSgenome.CSabaeus.UCSC.chlSab2_1.0.0.tar.gz", repos = NULL, type="source")
#BiocManager::install("motifmatchr")

library(chromVAR)
library(motifmatchr)
library(BSgenome.CSabaeus.UCSC.chlSab2)

library(plyr)
library(DESeq2)
library(ggplot2)
library(data.table)
library(biomaRt)
library(ggrepel)
library(pheatmap)
library(dplyr)
library(clusterProfiler)
library(stringr)
library(reactome.db)
library(OmnipathR)
library(gprofiler2)
library(igraph)
library(preprocessCore)
library(apeglm)
library(limma)

### helper function ####
substrLeft <- function(x, n){
  substr(x, 1, nchar(x)-n)
}

substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}
########################
```

# peak calling

```{r, eval=F}
####
#### Vero pInd20 Spike analysis
####

module load gcc/6.2.0
module load python/2.7.12
module load macs2/2.1.1.20160309
 
macs2 callpeak --nomodel --shift -100 --extsize 200 --nolambda -q 0.01 --keep-dup all --call-summits -t ../bowtie2/*_sorted.noM.norep.bam -g hs -n vero_pind20 --outdir ./

module load gcc/6.2.0 bedtools/2.27.1

cat vero_pind20_peaks.narrowPeak | awk '{print "chr"$0}' > vero_pind20_peaks.narrowPeak.bed

bedtools subtract -a vero_pind20_peaks.narrowPeak.bed -b /home/jdl22/pandolfi/myNGSdir/GRCh38/hg38-blacklist.v2.bed > vero_pind20_peaks_noblacklist.bed

cut -c 4- vero_pind20_peaks_noblacklist.bed > vero_pind20_peaks_noblacklist.nochr.bed

bedtools multicov -bams ../bowtie2/*_sorted.noM.norep.bam -bed vero_pind20_peaks_noblacklist.nochr.bed > vero_pind20_atacseq_count.txt
 
for f in ../bowtie2/*_sorted.noM.norep.bam; do echo $(basename $f _sorted.noM.norep.bam); done | tr '\n' '\t' | paste <(echo "position") - > vero_pind20_header.txt
 
cat vero_pind20_atacseq_count.txt | awk -vOFS='\t' '{printf "chr"$1":"$2"-"$3"\t"; for (i=11; i<NF; i++) printf $i"\t"; print $NF}' > vero_pind20_atacseq_count.tmp.txt
 
cat vero_pind20_header.txt vero_pind20_atacseq_count.tmp.txt > vero_pind20_atacseq_counts.txt

```

```{r, eval=T}
peakfile <- "datafiles/vero_pind20_peaks_noblacklist.clipped.nonzero.bed" # FDR 1% from macs2
bamfiles <- list.files(path="/Users/jonathanlee/Desktop/atacseq_veropind20/",pattern=".bam", full.names=T)
load("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/motifs/csab_cisbp_pwms.rda")
```

```{r, eval=T}
# read in ATAC-seq read pileups within peaks from bash
atac_counts <- read.table(file="datafiles/vero_pind20_atacseq_counts.txt", header=T)
atac_counts <- atac_counts[!duplicated(atac_counts),]
row.names(atac_counts) <- atac_counts[,1]
atac_counts <- atac_counts[,-1]

# read in ATAC-seq library information
sampleinfo <- read.csv(file="sample_info.csv")
row.names(sampleinfo) <- sampleinfo$Sample

colnames(atac_counts) <- str_replace_all(unlist(strsplit(colnames(atac_counts), "_"))[c(F,F,T)], "S", "")
sampleinfo <- sampleinfo[match(colnames(atac_counts), sampleinfo$Number),]
colnames(atac_counts) <- row.names(sampleinfo)

sampleinfo$Group <- factor(unlist(strsplit(sampleinfo$Sample, "-"))[c(F,F,T,F,F)], levels=c("EGFP", "Spike", "Alpha", "Delta"))
```

```{r, eval=T, fig.width=1, fig.height=0.8}
for(f in list.files(path="datafiles/", pattern="insertsize2.txt", full.names = T)){
  insertsize <- read.table(file=f, sep='\t', header=T)
  print(ggplot(insertsize, aes(x=size, y=count)) + geom_line() + theme_classic() + ylim(c(0,20000)) + xlim(c(0,1000)) + xlab("Insert Size (bp)") + ylab("# Reads") + ggtitle(subset(sampleinfo, Number==str_replace_all(strsplit(f, "_")[[1]][3], "S", ""))$Sample) + theme(plot.title = element_text(size = 10)))
}
```

```{r, eval=T, fig.width=1, fig.height=1}
tss_values <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="distToTSS.txt")){
  tss <- read.table(file=f, sep=' ', header=T)
  print(ggplot(tss, aes(x=Position/1000, y=Count)) + geom_line() + theme_classic() + ylim(c(0,2000)) + xlab("TSS Position (kb)") + ylab("# Reads")+ ggtitle(subset(sampleinfo, Number==str_replace_all(strsplit(f, "_")[[1]][3], "S", ""))$Sample) + theme(plot.title = element_text(size = 10))) 
  tss_values <- rbind(tss_values, cbind(sample=f, tss=mean(tss[c(2001 + seq(-5,5)),1])/mean(tss[1:200,1])))
}
```

```{r, eval=T}
tss_values <- data.frame(tss_values)
tss_values$TSS <- as.numeric(as.character(tss_values$tss))
tss_values$index <- str_replace_all(unlist(strsplit(tss_values$sample, "_"))[c(F,F,T,F)], "S", "")

tss_values <- merge(tss_values, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
tss_values <- tss_values[,c("Sample", "Group", "TSS")]
```

```{r, eval=T, fig.width=1.5, fig.height=0.8}
ggplot(tss_values, aes(x=Sample, y=TSS)) + geom_bar(stat="identity") + theme_classic() + geom_hline(yintercept=0) + coord_flip() + xlab("") + ylab("TSS-to-Background\nInsertion Ratio")

ggplot(tss_values, aes(x=Group, y=TSS)) + geom_point() + theme_classic() + geom_hline(yintercept=0, lty=3) + coord_flip() + xlab("") + ylab("TSS-to-Background\nInsertion Ratio")

```

```{r, eval=T, fig.width=1.5, fig.height=2}
nonmito_prop <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="mito.log")){
#for(f in list.files(path=".", pattern="mito.log")){
  nonmito_prop <- rbind(nonmito_prop, cbind(sample=f, prop=as.character(tail(read.table(file=f),1)$V5)))
}

nonmito_prop <- data.frame(nonmito_prop)
nonmito_prop$prop <- as.numeric(as.character(nonmito_prop$prop))

nonmito_prop$index <- str_replace_all(unlist(strsplit(nonmito_prop$sample, "_"))[c(F,F,T,F)], "S", "")
nonmito_prop <- merge(nonmito_prop, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
nonmito_prop <- nonmito_prop[,c("Sample", "Group", "prop")]

ggplot(nonmito_prop, aes(x=Sample, y=prop)) + geom_point() + ylim(c(0,1.1)) + theme_classic() + xlab("") + ylab("Non-Mitochondrial\nRead Proportion") + coord_flip()
```

```{r, eval=T, fig.width=2, fig.height=2}
peakanno <- NULL
for(f in list.files(path="datafiles/", full.names = T, pattern="5000000.annstats.txt")){
#for(f in list.files(path=".", pattern="5000000.annstats.txt")){
  peakanno <- rbind(peakanno, cbind(sample=f, anno=as.character(head(read.table(file=f, sep='\t', header=T),5)[,1]), ratio=as.numeric(as.character(head(read.table(file=f, sep='\t', header=T),5)[,4]))))
}

peakanno <- data.frame(peakanno)
peakanno$ratio <- as.numeric(as.character(peakanno$ratio))

peakanno$index <- str_replace_all(unlist(strsplit(peakanno$sample, "_"))[c(F,F,T,F)], "S", "")
peakanno <- merge(peakanno, sampleinfo[,c("Number", "Sample", "Group")], by.x="index", by.y="Number")
peakanno <- peakanno[,c("Sample", "Group", "anno", "ratio")]

ggplot(peakanno, aes(x=anno, y=ratio, fill=Sample)) + geom_bar(stat="identity", position="dodge") + theme_classic() + ylab("Log2 Fold Change") + xlab("Annotation") + geom_hline(yintercept=0, lty=3) + coord_flip() + theme(legend.position="none")
```

```{r, eval=T}
atac_dds <- DESeqDataSetFromMatrix(countData = atac_counts, colData = sampleinfo, design = ~Group)
atac_dds$Group <- relevel(atac_dds$Group, "EGFP")
atac_dds <- estimateSizeFactors(atac_dds)
```

```{r, eval=T}
atac_vst <- assay(varianceStabilizingTransformation(atac_dds))
```

```{r, eval=T, fig.width=2, fig.height=1.2}
atac_pc <- prcomp(t(atac_vst))

ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst),"Group"]), aes(PC1, PC2)) + 
  geom_point(aes(colour = Sample), size=3)  + theme_classic() + xlab(paste0("PC1 (", 100*summary(atac_pc)$importance[2,1], "% of variance)")) + ylab(paste0("PC2 (", 100*summary(atac_pc)$importance[2,2], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red")

ggplot(data.frame(atac_pc$x, Sample=sampleinfo[colnames(atac_vst),"Group"]), aes(PC3, PC4)) + 
  geom_point(aes(colour = Sample), size=3)  + theme_classic() + xlab(paste0("PC3 (", 100*summary(atac_pc)$importance[2,3], "% of variance)")) + ylab(paste0("PC4 (", 100*summary(atac_pc)$importance[2,4], "% of variance)")) + geom_vline(xintercept=0, lty=3, color="red")
```

```{r, eval=T}
atac_dds <- DESeq(atac_dds)
```

```{r, eval=F}
atac_sigpeaks <- data.frame()
for(l in levels(sampleinfo$Group)[-1]){
  atac_sigpeaks <- rbind(atac_sigpeaks, data.frame(group = l,
        sigpeaks = -dim(subset(results(atac_dds, contrast=c("Group", l, "EGFP")), padj < 0.05 & log2FoldChange < 0))[1],
        sign = "neg"))
  atac_sigpeaks <- rbind(atac_sigpeaks, data.frame(group = l,
        sigpeaks = dim(subset(results(atac_dds, contrast=c("Group", l, "EGFP")), padj < 0.05 & log2FoldChange > 0))[1],
        sign = "pos"))
}
```

```{r, eval=T, fig.width=.7, fig.height=1.2}
ggplot(atac_sigpeaks, aes(x=group, y=sigpeaks, color=sign)) + geom_point() + ylim(c(-6000, 6000)) + theme_classic() + geom_hline(yintercept=0) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant Peaks\n(vs. EGFP)") + xlab("")
```

```{r, eval=T, fig.width=1.5, fig.height=1.5}
vero_spike_results <- data.frame(lfcShrink(dds=atac_dds, coef=2)) #contrast=c("Group", "Spike", "EGFP")))

ggplot(vero_spike_results, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(subset(vero_spike_results), padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(subset(vero_spike_results), padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(subset(vero_spike_results), padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-3.5,3.5)) + ylim(c(0,45)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nVero Spike vs. EGFP ATAC-seq")

vero_alpha_results <- data.frame(lfcShrink(dds=atac_dds, coef=3))

ggplot(vero_alpha_results, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(subset(vero_alpha_results), padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(subset(vero_alpha_results), padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(subset(vero_alpha_results), padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-3.5,3.5)) + ylim(c(0,45)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nVero Alpha vs. EGFP ATAC-seq")

vero_delta_results <- data.frame(lfcShrink(dds=atac_dds, coef=4))

ggplot(vero_delta_results, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(subset(vero_delta_results), padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(subset(vero_delta_results), padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(subset(vero_delta_results), padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-3.5,3.5)) + ylim(c(0,45)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nVero Delta vs. EGFP ATAC-seq")
```

```{r, eval=T}
atac_results_vero_nut <- read.table(file="../atacseq_vero_nut/atac_results_vero_nut.txt", header=T)
```


```{r, eval=T}
cor((subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))

cor((subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))

cor((subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))
```

```{r, eval=T}
cor.test((subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))

cor.test((subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))

cor.test((subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))
```

```{r, eval=T}
cor_delta <- cor.test((subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))

cor_alpha <- cor.test((subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))

cor_spike <- cor.test((subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.x),
    (subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < .2 & padj.y < .2)$log2FoldChange.y))
```

```{r, eval=T}
cor_df <- data.frame(voc = factor(c("Spike", "Alpha", "Delta"), levels=rev(c("Spike", "Alpha", "Delta"))),
                     cor = c(cor_spike$estimate, cor_alpha$estimate, cor_delta$estimate), 
                     pval = c(cor_spike$p.value, cor_alpha$p.value, cor_delta$p.value))
```

```{r, eval=T, fig.width=1.8, fig.height=1.2}
ggplot(cor_df, aes(y = cor, x = voc)) +
  geom_segment(yend=0, aes(xend=voc), lty=3) + theme_classic() + ylab("Pearson Correlation\nvs. Nutlin-3a Treatment\nATAC-seq") + 
  geom_point(aes(color=(pval)))  +
  xlab("") + coord_flip() + 
  theme(legend.position = "right", axis.text.y = element_text(vjust=0.5)) +
  scale_color_continuous(low="red", high="blue", name="p-value") +
  geom_hline(yintercept=0, lwd=0.1) + ylim(c(0,1))
```

```{r, eval=T, fig.width=1.5, fig.height=1.5}
ggplot(subset(merge(atac_results_vero_nut, vero_delta_results, by=0), padj.x < 0.2 & padj.y < 0.2)) + geom_point(aes(x=(log2FoldChange.x), y=(log2FoldChange.y)), alpha=0.5) + xlim(c(-3.5,3.5)) + ylim(c(-3.5,3.5)) + geom_smooth(method="lm", formula=y~x, aes(x=(log2FoldChange.x), y=(log2FoldChange.y))) + theme_classic() + geom_hline(yintercept=0, lty=3) + geom_vline(xintercept=0, lty=3) + xlab("Vero Nutlin-3a vs. DMSO\nLog2 Fold Change, ATAC-seq") + ylab("Vero Delta vs. EGFP\nLog2 Fold Change, ATAC-seq")

ggplot(subset(merge(atac_results_vero_nut, vero_alpha_results, by=0), padj.x < 0.2 & padj.y < 0.2)) + geom_point(aes(x=(log2FoldChange.x), y=(log2FoldChange.y)), alpha=0.5) + xlim(c(-3.5,3.5)) + ylim(c(-3.5,3.5)) + geom_smooth(method="lm", formula=y~x, aes(x=(log2FoldChange.x), y=(log2FoldChange.y))) + theme_classic() + geom_hline(yintercept=0, lty=3) + geom_vline(xintercept=0, lty=3) + xlab("Vero Nutlin-3a vs. DMSO\nLog2 Fold Change, ATAC-seq") + ylab("Vero Alpha vs. EGFP\nLog2 Fold Change, ATAC-seq")

ggplot(subset(merge(atac_results_vero_nut, vero_spike_results, by=0), padj.x < 0.2 & padj.y < 0.2)) + geom_point(aes(x=(log2FoldChange.x), y=(log2FoldChange.y)), alpha=0.5) + xlim(c(-3.5,3.5)) + ylim(c(-3.5,3.5)) + geom_smooth(method="lm", formula=y~x, aes(x=(log2FoldChange.x), y=(log2FoldChange.y))) + theme_classic() + geom_hline(yintercept=0, lty=3) + geom_vline(xintercept=0, lty=3) + xlab("Vero Nutlin-3a vs. DMSO\nLog2 Fold Change, ATAC-seq") + ylab("Vero Spike vs. EGFP\nLog2 Fold Change, ATAC-seq")
```

```{r, eval=T}
write.table(vero_spike_results, file="atacseq_vero_spike.txt", sep='\t', quote=F)
write.table(vero_alpha_results, file="atacseq_vero_alpha.txt", sep='\t', quote=F)
write.table(vero_delta_results, file="atacseq_vero_delta.txt", sep='\t', quote=F)
```



