---
title: "2021-01-07_atacseq"
output: html_document
---

```{r setup, include=FALSE}
#devtools::install_github("GreenleafLab/chromVARmotifs")
#source("http://bioconductor.org/biocLite.R")
#biocLite("chromVAR")
#biocLite("motifmatchr")
#biocLite("BSgenome.Hsapiens.UCSC.hg38")
#BiocManager::install("chromVAR")
#install.packages("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/chlsab_bsgenome/BSgenome.CSabaeus.UCSC.chlSab2_1.0.0.tar.gz", repos = NULL, type="source")
#BiocManager::install("motifmatchr")

#library(chromVAR)
#library(motifmatchr)
#library(BSgenome.Hsapiens.UCSC.hg38)

library(plyr)
library(DESeq2)
library(ggplot2)
library(data.table)
library(biomaRt)
library(ggrepel)
library(pheatmap)
library(dplyr)
#library(clusterProfiler)
library(stringr)
library(reactome.db)
#library(OmnipathR)
#library(gprofiler2)
library(igraph)
library(preprocessCore)
#library(apeglm)
library(limma)

### helper function ####
substrLeft <- function(x, n){
  substr(x, 1, nchar(x)-n)
}

substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}
########################
```

# peak calling

```{r, eval=F}
####
#### A549 PIND20 analysis
####

module load gcc/6.2.0
module load python/2.7.12
module load macs2/2.1.1.20160309
module load bedtools/2.27.1

macs2 callpeak --nomodel --shift -100 --extsize 200 --nolambda -q 0.01 --keep-dup all --call-summits -t ../bowtie2/*_sorted.noM.norep.bam -g hs -n a549pind20 --outdir ./


cat a549pind20_peaks.narrowPeak | awk '{print "chr"$0}' > a549pind20_peaks.narrowPeak.bed

bedtools subtract -a a549pind20_peaks.narrowPeak.bed -b /home/jdl22/pandolfi/myNGSdir/GRCh38/hg38-blacklist.v2.bed > a549pind20_peaks_noblacklist.bed

cut -c 4- a549pind20_peaks_noblacklist.bed > a549pind20_peaks_noblacklist.nochr.bed

bedtools multicov -bams ../bowtie2/*_sorted.noM.norep.bam -bed a549pind20_peaks_noblacklist.nochr.bed > a549pind20_atacseq_count.txt
 
for f in ../bowtie2/*_sorted.noM.norep.bam; do echo $(basename $f _sorted.noM.norep.bam); done | tr '\n' '\t' | paste <(echo "position") - > a549pind20_header.txt
 
cat a549pind20_atacseq_count.txt | awk -vOFS='\t' '{printf "chr"$1":"$2"-"$3"\t"; for (i=11; i<NF; i++) printf $i"\t"; print $NF}' > a549pind20_atacseq_count.tmp.txt
 
cat a549pind20_header.txt a549pind20_atacseq_count.tmp.txt > a549pind20_atacseq_counts.txt

```

```{r, eval=T}
#peakfile <- "datafiles/a549pind20_peaks_noblacklist.bed" # FDR 1% from macs2
#bamfiles <- list.files(path="/Users/jonathanlee/Desktop/atacseq_a549pind20/",pattern=".bam", full.names=T)
#load("/Users/jonathanlee/Dropbox/RESEARCH/RESEARCH_SLACK/P COVID iDAPT/motifs/hs_cisbp_pwms.rda")
```

```{r, eval=T}
# read in ATAC-seq read pileups within peaks from bash
atac_counts <- read.table(file="datafiles/a549ace2_pind_atacseq_counts.txt", header=T)
atac_counts <- atac_counts[!duplicated(atac_counts),]
row.names(atac_counts) <- atac_counts[,1]
atac_counts <- atac_counts[,-1]

# read in ATAC-seq library information
sampleinfo <- data.frame(Group = c("EGFP", "Spike", "Alpha", "Delta",
                                   "EGFP", "Spike", "Alpha", "Delta",
                                   "EGFP", "Spike", "Alpha", "Delta",
                                   "EGFP", "Spike", "Alpha", "Delta",
                                   "DMSO", "DMSO", "Nutlin", "Nutlin"),
                         Geno = c(rep("WT",8), rep("KO",8), rep("TX",4)), 
                            row.names = c("A549ACE2-EGFP-1",
                                          "A549ACE2-Spike-1",
                                          "A549ACE2-Alpha-1",
                                          "A549ACE2-Delta-1",
                                          "A549ACE2-EGFP-2",
                                          "A549ACE2-Spike-2",
                                          "A549ACE2-Alpha-2",
                                          "A549ACE2-Delta-2",
                                          "A549ACE2-KO-EGFP-1",
                                          "A549ACE2-KO-Spike-1",
                                          "A549ACE2-KO-Alpha-1",
                                          "A549ACE2-KO-Delta-1",
                                          "A549ACE2-KO-EGFP-2",
                                          "A549ACE2-KO-Spike-2",
                                          "A549ACE2-KO-Alpha-2",
                                          "A549ACE2-KO-Delta-2",
                                          "A549-DMSO-1",
                                          "A549-DMSO-2",
                                          "A549-Nutlin-1",
                                          "A549-Nutlin-2"))

colnames(atac_counts) <- row.names(sampleinfo)
```

```{r, eval=T}
sampleinfo <- subset(sampleinfo, Group != "Alpha" & Geno != "TX")
sampleinfo$Sample <- paste(sampleinfo$Group, sampleinfo$Geno)
atac_counts <- atac_counts[,row.names(sampleinfo)]
```

```{r, eval=T}
set.seed(1)
atac_dds <- DESeqDataSetFromMatrix(countData = atac_counts, colData = sampleinfo, design = ~Sample)
atac_dds$Sample <- relevel(atac_dds$Sample, "EGFP WT")
atac_dds <- DESeq(atac_dds)

atac_results_spike <- data.frame(lfcShrink(dds=atac_dds, coef=6))
atac_results_delta <- data.frame(lfcShrink(dds=atac_dds, coef=3))

atac_dds_ko <- DESeqDataSetFromMatrix(countData = atac_counts, colData = sampleinfo, design = ~Sample)
atac_dds_ko$Sample <- relevel(atac_dds_ko$Sample, "EGFP KO")
atac_dds_ko <- DESeq(atac_dds_ko)

atac_results_spike_ko <- data.frame(lfcShrink(dds=atac_dds_ko, coef=5))
atac_results_delta_ko <- data.frame(lfcShrink(dds=atac_dds_ko, coef=2))
```

```{r, eval=T, fig.width=1.5, fig.height=1.5}
ggplot(atac_results_spike, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(atac_results_spike, padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(atac_results_spike, padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(atac_results_spike, padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-6,6)) + ylim(c(0,15)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nA549ACE2\nSpike vs. EGFP ATAC-seq")

ggplot(atac_results_delta, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(atac_results_delta, padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(atac_results_delta, padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(atac_results_delta, padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-6,6)) + ylim(c(0,15)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nA549ACE2\nDelta vs. EGFP ATAC-seq")
```

```{r, eval=T, fig.width=1.5, fig.height=1.5}
ggplot(atac_results_spike_ko, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(atac_results_spike_ko, padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(atac_results_spike_ko, padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(atac_results_spike_ko, padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-6,6)) + ylim(c(0,15)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nA549ACE2 TP53ko\nSpike vs. EGFP ATAC-seq")

ggplot(atac_results_delta_ko, aes(x=log2FoldChange, y=-log10(padj))) + geom_point(pch=20, color="lightgray", alpha=0.5, data=subset(atac_results_delta_ko, padj > 0.05)) + geom_point(pch=20, color="skyblue", data=subset(atac_results_delta_ko, padj < 0.05 & log2FoldChange > 0), alpha=0.5) + geom_point(pch=20, color="indianred", data=subset(atac_results_delta_ko, padj < 0.05 & log2FoldChange < 0), alpha=0.5) + theme_classic() + geom_hline(yintercept=0, lty=3) + xlim(c(-6,6)) + ylim(c(0,15)) + 
  geom_hline(yintercept=-log10(0.05), lty=3, color="red") + geom_vline(xintercept=0, lty=3, color="red") + ylab("-Log10 FDR") + xlab("Log2 Fold Change,\nA549ACE2 TP53ko\nDelta vs. EGFP ATAC-seq")
```

```{r, eval=T}
dim(subset(atac_results_spike, padj < 0.05 & log2FoldChange < 0))[1]
dim(subset(atac_results_spike, padj < 0.05 & log2FoldChange > 0))[1]
dim(subset(atac_results_delta, padj < 0.05 & log2FoldChange < 0))[1]
dim(subset(atac_results_delta, padj < 0.05 & log2FoldChange > 0))[1]
dim(subset(atac_results_spike_ko, padj < 0.05 & log2FoldChange < 0))[1]
dim(subset(atac_results_spike_ko, padj < 0.05 & log2FoldChange > 0))[1]
dim(subset(atac_results_delta_ko, padj < 0.05 & log2FoldChange < 0))[1]
dim(subset(atac_results_delta_ko, padj < 0.05 & log2FoldChange > 0))[1]
```

```{r, eval=T}
atac_sigpeaks <- data.frame(
  group = c("Spike", "Spike", "Delta", "Delta", "Spike KO", "Spike KO", "Delta KO", "Delta KO"),
  sigpeaks = c(-dim(subset(atac_results_spike, padj < 0.05 & log2FoldChange < 0))[1],
               dim(subset(atac_results_spike, padj < 0.05 & log2FoldChange > 0))[1],
               -dim(subset(atac_results_delta, padj < 0.05 & log2FoldChange < 0))[1],
               dim(subset(atac_results_delta, padj < 0.05 & log2FoldChange > 0))[1],
               -dim(subset(atac_results_spike_ko, padj < 0.05 & log2FoldChange < 0))[1],
               dim(subset(atac_results_spike_ko, padj < 0.05 & log2FoldChange > 0))[1],
               -dim(subset(atac_results_delta_ko, padj < 0.05 & log2FoldChange < 0))[1],
               dim(subset(atac_results_delta_ko, padj < 0.05 & log2FoldChange > 0))[1]),
  sign = rep(c("neg", "pos"), 4)
)

atac_sigpeaks$group <- factor(atac_sigpeaks$group, levels=unique(atac_sigpeaks$group))
```

```{r, eval=T, fig.width=.8, fig.height=1.5}
ggplot(atac_sigpeaks, aes(x=group, y=sigpeaks, color=sign, pch=sign)) + geom_point() + ylim(c(-6000, 6000)) + theme_classic() + geom_hline(yintercept=0) + scale_color_manual(values=c("red", "black")) + scale_shape_manual(values=c(1, 2)) + theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("# Significant Peaks\n(vs. EGFP)") + xlab("")
```

```{r, eval=T}
write.table(atac_results_spike, file="atac_a549ace2_spike.txt", sep='\t', quote=F)
write.table(atac_results_delta, file="atac_a549ace2_delta.txt", sep='\t', quote=F)
write.table(atac_results_spike_ko, file="atac_a549ace2_ko_spike.txt", sep='\t', quote=F)
write.table(atac_results_delta_ko, file="atac_a549ace2_ko_delta.txt", sep='\t', quote=F)
```


